File: Week_09/advanced_dp/91_decode_ways.java
class Solution {
    public int numDecodings(String s) {
        int len = s.length();
        int[] dp = new int[len+1];

        dp[0] = 1;
        dp[1] = s.charAt(0) == '0' ? 0 : 1;

        for (int i = 2; i < len + 1; i ++){
            int oneNum = 1;
            int twoNum = 1;
            int singleDigit = Integer.parseInt(s.substring(i-1, i));
            String doubleDigitStr = s.substring(i-2, i);
            int doubleDigit = Integer.parseInt(doubleDigitStr);

            if (singleDigit > 26 || singleDigit < 1)
                oneNum = 0;
            if (doubleDigitStr.charAt(0) == '0')
                twoNum = 0;
            if (doubleDigit > 26 || doubleDigit < 1)
                twoNum = 0;

            dp[i] = oneNum * dp[i - 1] + twoNum * dp[i - 2];
        }

        return dp[len];
    }
}

File: Week_03/python/91_decode_ways.py
class Solution:
    def numDecodings(self, s: str) -> int:
        if not s or s[0] == "0": return 0

        dp = [[0] * 2 for _ in range(len(s))]

        for i in range(len(s)):
            if i == 0:
                dp[0][0] = 1
                dp[0][1] = 0
                continue

            single = int(s[i])
            double = int(s[i - 1:i + 1])

            if double == 0:
                dp[i][0] = 0
                dp[i][1] = 0

            if double in range(1, 10):
                dp[i][0] = dp[i - 1][1]
                dp[i][1] = 0

            if double in [10, 20]:
                dp[i][0] = 0
                dp[i][1] = dp[i - 1][0]

            if double in list(range(11, 20)) + list(range(21, 27)):
                dp[i][0] = dp[i - 1][1] + dp[i - 1][0]
                dp[i][1] = dp[i - 1][0]

            if double >= 27:
                dp[i][0] = dp[i - 1][1] + dp[i - 1][0] if single != 0 else 0
                dp[i][1] = 0

        return sum(dp[-1])
